core/primitives/src/views.rs

```diff
@@ -1265,73 +1265,102 @@ impl Default for ExecutionMetadataView {
+                        format!("{:?}", ext_cost).to_ascii_uppercase(),
+                        profile_data.get_ext_cost(ext_cost),
+                    ));
+                }
 
-                // The order doesn't really matter, but the default one is just
-                // historical, which is especially unintuitive, so let's sort
-                // lexicographically.
-                //
-                // Can't `sort_by_key` here because lifetime inference in
-                // closures is limited.
-                costs.sort_by(|lhs, rhs| {
-                    lhs.cost_category.cmp(&rhs.cost_category).then(lhs.cost.cmp(&rhs.cost))
-                });
+                Some(costs)
+            }
+            ExecutionMetadata::V3(profile) => {
+                // Add actions, wasm op, and ext costs in groups.
+                // actions costs are 1-to-1
+                let mut costs: Vec<CostGasUsed> = ActionCosts::iter()
+                    .flat_map(|cost| {
+                        let gas_used = profile.get_action_cost(cost);
+                        (gas_used > 0).then(|| {
+                            CostGasUsed::action(
+                                format!("{:?}", cost).to_ascii_uppercase(),
+                                gas_used,
+                            )
+                        })
+                    })
+                    .collect();
+
+                // wasm op is a single cost, for historical reasons it is inaccurately displayed as "wasm host"
+                let wasm_gas_used = profile.get_wasm_cost();
+                if wasm_gas_used > 0 {
+                    costs.push(CostGasUsed::wasm_host(
+                        "WASM_INSTRUCTION".to_string(),
+                        wasm_gas_used,
+                    ));
+                }
 
-                // need to remove duplicate entries due to cost conflation
-                costs.dedup();
+                // ext costs are 1-to-1
+                for ext_cost in ExtCosts::iter() {
+                    let gas_used = profile.get_ext_cost(ext_cost);
+                    if gas_used > 0 {
+                        costs.push(CostGasUsed::wasm_host(
+                            format!("{:?}", ext_cost).to_ascii_uppercase(),
+                            gas_used,
+                        ));
+                    }
+                }
 
                 Some(costs)
             }
         };
-        ExecutionMetadataView { version: 1, gas_profile }
+        if let Some(ref mut costs) = gas_profile {
+            // The order doesn't really matter, but the default one is just
+            // historical, which is especially unintuitive, so let's sort
+            // lexicographically.
+            //
+            // Can't `sort_by_key` here because lifetime inference in
+            // closures is limited.
+            costs.sort_by(|lhs, rhs| {
+                lhs.cost_category.cmp(&rhs.cost_category).then(lhs.cost.cmp(&rhs.cost))
```

Comment:
I suspect clippy would complain about the `then` here, possibly saying something about it being executed regardless of whether the first check passed or not. I suspect the compiler optimizes this out well either way, though…